TST: use Hypothesis for property-based tests

mdhaber opened the pull request:
#### Reference issue
gh-12197
Toward gh-15928

#### What does this implement/fix?
This adds [Hypothesis](https://hypothesis.readthedocs.io/en/latest/index.html) as a testing dependency and uses it in a property-based test of `scipy.stats.ttest_1samp`. The test takes approximately 3s to complete locally, so I'll mark it `xslow` after confirming that it is passing consistently on CI.

#### Additional information
I defined a custom strategy to filter out examples that cause RuntimeWarnings and failures that aren't the fault of the function being tested. An alternative strategy would have been to change the assertions that are made when the data is atypical, but the original purpose was to test the CI/p-value relationship, not edge case handling.

Changes to files other than `test_stats.py` and `.gitignore` are guesses based on the presence of `pytest` nearby. Please let me know what is extraneous and what is missing.

Must discuss on mailing list before merging.

Questions (and answers) about Hypothesis:
- ~~GitHub actions jobs are failing; looks Hypothesis didn't install on these platforms. What else do I need to change?~~ Needed to look more closely for appearances of `pytest`.
- ~~I still need to investigate how we will [reproduce failures](https://hypothesis.readthedocs.io/en/latest/reproducing.html) that occur on CI. If you know, please post.~~ Hypothesis prints decorators in the log that can be used to reproduce failures locally.
- ~~When there is a failure (usually due to a mistake in configuration), the test takes a *long* time time to finish. Why?~~ Hypothesis tries to find a **M**RE.
- ~~Is there a way to limit the approximate duration of the test? This takes over a second. That's a lot more time than a property -based test that I might write, even if it's heavily parametrized. What is it doing?~~ There is, See [settings](https://hypothesis.readthedocs.io/en/latest/settings.html) like `max_examples` and `deadline`.
- ~~I tried to use `hypothesis.extra.numpy.arrays` to generate the test data, but ran into trouble.~~
  - ~~`from_dtype` documentation states "This is particularly useful when kwargs are passed through from `arrays()` which allow a variety of numeric dtypes, as it seamlessly handles the width or representable bounds for you." How do I do this?~~ instead of passing into `arrays` the object returned by `from_dtype`, pass a dict with the kwargs into `arrays`. It will pass them into `from_dtype`.
  - What is the difference between `width` and `sizes`?

tupui reviewed (commented):
Cool! I will play with that locally (I actually never used hypothesis for real.)

For CI I would suppose we could add a marker and not run that always. I would expect that these take time?

mdhaber commented:
My impression is that Hypothesis shines at testing edge cases, but I'm not giving it much flexibility to do that here. So depending on what it is doing, it might be ok to reduce the number of cases it tries. Otherwise, yes, this would need to be marked slow or xslow.

mdhaber commented:
@tirthasheshpatel you worked with Hypothesis before. Do you have answers to any of the questions in the top post?

tylerjereddy commented on scipy/stats/tests/test_stats.py:
```diff
@@ -3369,6 +3371,29 @@ def test_1samp_ci_iv(self):
+        popmean = l if alternative == 'greater' else u
+        popmean = np.expand_dims(popmean, axis=axis)
+        res = stats.ttest_1samp(data, popmean, alternative=alternative, axis=axis)
+        np.testing.assert_allclose(res.pvalue, 1 - alpha)
```
I advocated a bit for the adoption of `hypothesis` in NumPy a few years ago, so wanted to mention a few things. First, I believe @Zac-HD was helpful in getting NumPy up to speed with `hypothesis` infrastructure in the early stages (as a maintainer of `hypothesis` proper).

> I still need to investigate how we will [reproduce failures](https://hypothesis.readthedocs.io/en/latest/reproducing.html) that occur on CI. If you know, please post.

It probably depends a bit on the type of failure, but for example I intentionally broke your test locally and `hypothesis` just gave me the parameters leading to failure.

```
E   Falsifying example: test_pvalue_ci(
E       self=<scipy.stats.tests.test_stats.TestStudentTest object at 0x7fbd06ad4090>,
E       alternative='greater',
E       shape=(2,),
E       alpha=0.5,
E   )

```

I'm almost certain there's another failure scenario, but it is also dealt with by the docs you linked to, and I've seen it a bunch of times over the years--adding a decorator to reproduce tricky cases:

`You can reproduce this example by temporarily adding @reproduce_failure(..., b'AAAA//AAAAAAAAEA') as a decorator on your test case`

> Is there a way to limit the approximate duration of the test? This takes over a second. That's a lot more time than a property -based test that I might write, even if it's heavily parametrized. What is it doing?

Zac will know better than I, but IIRC test exploration will cycle to extreme values and kind of expand and contract over that space a bit, and it may also be the case that `hypothesis` explores inputs that result in unexpectedly slow performance as well (which may also be useful info). I know there are a bunch of ways to limit where `hypothesis` stands on the spectrum between "quick check of a few cases" and "full blown fuzzer," but approaches like example count tuning  `@settings(max_examples=50)` (instead of default `100`) or the `deadline` in milliseconds, etc. That said, I believe the default settings are intended for regular CI flushing.

I suppose we could `grep` the NumPy code base a bit--I see one test with `@settings(max_examples=10000, deadline=2000)` to control its search expansion for example.

Another useful thing is skipping cases that you might know are very time consuming and/or irrelevant with stuff like `assume(not isnan(x))` if `nan` inputs are problematic/slow but also irrelevant to your test.

Zac-HD commented:
👋 I'm always happy to help foundational projects improve their testing, so feel free to tag me on your PRs... although I'm camping and mostly-offline for the next week and a half, so please excuse delays.

-------------

### Shrinking, and why your failing test is slow
Failing tests typically take a bit longer because once we find a failing example, Hypothesis will search from a _minimal_ failing example by trying local variations (with momentum-like tricks to speed this up).

This specific case is almost pessimal, because of the use of a non-Hypothesis-integrated RNG: there's basically no structure that we can use to shrink more efficiently, and making any local move invalidates most of the information we have so far.  Fortunately, generating the array with Hypothesis should help with this; I'll show that below.

### Reproducing failures and the database
As Tyler notes, Hypothesis will print the specific failing inputs; and if you use the [`print_blob=True` setting](https://hypothesis.readthedocs.io/en/latest/settings.html#hypothesis.settings.print_blob) (which defaults-on in CI) we'll also print a decorator which will replay the failing example.  [See here for more details](https://hypothesis.readthedocs.io/en/latest/reproducing.html).

Failing examples are also added to [the Hypothesis database](https://hypothesis.readthedocs.io/en/latest/database.html), so locally you can replay failing examples just be re-running your tests.  From CI, you can also upload the database as an artifact and then automatically pull it down with our `GitHubArtifactDatabase` - this is a newer workflow, but [you can see the artifact config here](https://github.com/HypothesisWorks/hypothesis/blob/master/.github/workflows/fuzz.yml) (in our experimental fuzzing setup).


### About `npst.arrays()`, `npst.from_dtype()`, and some-distinct-elements

`npst.arrays(dtype=dtype, shape=..., elements=dict(min_value=..., max_value=...)).filter(has_more_than_one_distinct_element)`  should do the trick here I think - you'll need to define the filter predicate appropriately, but rejecting arrays with a single distinct element should perform a bit better (and check some more edge cases) than only unique arrays.

mdhaber commented:
Thanks @tylerjereddy @Zac-HD. All of this was very helpful.

Now the test configuration looks like:

```python3
    shape = npst.array_shapes(min_dims=1, min_side=2)
    alpha = strategies.floats(1e-8, 1-1e-8, allow_nan=False)
    data = npst.arrays(dtype=npst.floating_dtypes(), 
                       shape=shape).filter(_moment_warns)
    @given(data=data, alpha=alpha)
```

The filter eliminates cases in which the variance is zero or its calculation raises a warning, which is usually about precision loss due to catastrophic cancellation. We can't expect an accurate t-test if the variance calculation is problematic.

However, because the filter doesn't know what the `axis` is going to be, it currently tries all possible axes:

```python3
def _moment_warns(data):
    with warnings.catch_warnings():
        warnings.simplefilter("error")
        for axis in range(len(data.shape)):
            try:
                var = stats.moment(data, moment=2, axis=axis)
                assert var > 0
                return True
            except Exception:
                return False
```

I considered a few other possibilities. Instead of using `filter`, I could call `_moment_warns` inside the test after `axis` is chosen. If `_moment_warns(data)` is True, I could simply `return` without making assertions. However, if Hypothesis doesn't know when tests are rejected, it's possible that it could try *only* cases that are rejected, and then we haven't really tested anything. Additional possibilities when `_moment_warns(data)` is True are to change the assertion (e.g. make sure the t-test fails gracefully) or to somehow perturb the `data` to avoid the problem in variance calculation.

What would you suggest?

Follow-up question for @tylerjereddy: do you know where else I should add `hypothesis` so that it gets installed on GitHub actions?

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -3368,6 +3382,23 @@ def test_1samp_ci_iv(self):
 
+    shape = npst.array_shapes(min_dims=1, min_side=2)
+    alpha = strategies.floats(1e-8, 1-1e-8, allow_nan=False)
+    data = npst.arrays(dtype=npst.floating_dtypes(),
```
```suggestion
    elements = dict(allow_nan=False, allow_infinity=False)
    data = npst.arrays(dtype=npst.floating_dtypes(), elements=elements,```

Zac-HD replied:
I recommend spending your time thinking about the domain of possible inputs rather than the distribution, which in this case means leaving subnormals enabled.

The distribution is in any case a very complicated mixture of heuristics, feedback-guided search, replaying verbatim or modified examples from earlier, etc.

Zac-HD commented:
It's idiomatic to define the strategies you're using inside the call to `@given()`, unless you intend to use them in multiple tests - in that case I'd write them as module-globals, or define a helper function which returns the strategy you want (also a good way to add some customization).

Hypothesis also has [the `assume()` function](https://hypothesis.readthedocs.io/en/latest/details.html#making-assumptions), which implements "filter, but inside your test".  In cases where you can easily use filter I prefer that, because it allows for finer-grained retries, but this sounds like a perfect case for assume.

mdhaber commented:
> It's idiomatic to define the strategies you're using inside the call to `@given()`

Like:

```python3
    @given(
        alpha=strategies.floats(1e-8, 1-1e-8, allow_nan=False),
        data=npst.arrays(
            dtype=npst.floating_dtypes(),
            elements=dict(allow_nan=False, allow_infinity=False),
            shape=npst.array_shapes(min_dims=1, min_side=2)
        ).filter(_moment_warns))
```

?
I prefer to avoid nested line continuations. If I want to "define a helper function", would it be ok to make it a helper static method so it can be collocated with the test, e.g.

```python3
    @staticmethod
    def pvalue_ci_test_strategy():
        alpha = strategies.floats(1e-8, 1-1e-8, allow_nan=False)
        dtype = npst.floating_dtypes()
        elements = dict(allow_nan=False, allow_infinity=False)
        shape = npst.array_shapes(min_dims=1, min_side=2)
        data = npst.arrays(dtype=dtype, elements=elements, shape=shape)
        return dict(alpha=alpha, data=data.filter(_moment_warns))
```

> Hypothesis also has [the assume() function](https://hypothesis.readthedocs.io/en/latest/details.html#making-assumptions),

That looks perfect.

Is there any way for `hypothesis` to generate an `axis` that is compatible with the shape of `data`, or do I need to do that inside the test (as I am)?

Zac-HD commented:
Idiom: yep, that's what I meant.  If you prefer an alternative style it's your code of course!  A helper staticmethod should indeed work just fine, though if you want to use [`@st.composite`](https://hypothesis.readthedocs.io/en/latest/data.html#composite-strategies) I don't recall offhand what order of decorators you'd need.

To fix the health-check warning - you can also just suppress warnings, but better to fix them if that's practical - I'd try to compute the axes for which this should work, `assume(valid_axes)` (I guess you could filter on "there exists a valid axis" too), and then `data.draw(st.sampled_from(valid_axes))`.

Expanding on this a bit, you've hit the complicated kind of data-generation pretty early: cases where you need arguments that have a particular relationship to each other, which are awkward or infeasible to express as separate strategies.  When this happens, you'll want to be able to pick a value from one strategy, then use that value to define the strategy from which you draw a second value (and so on).  The two ways to do this are with [`@st.composite` to define a custom strategy function](https://hypothesis.readthedocs.io/en/latest/data.html#composite-strategies), or using [`st.data()` to draw data inside your _test function_](https://hypothesis.readthedocs.io/en/latest/data.html#drawing-interactively-in-tests).  They're equivalent in power: `st.data()` has the advantage of not needing a second function, can _also_ be based on the behaviour of your system-under-test so far (not relevant here), and disadvantage that you can't use it with `@example()` and the printed reproduction instructions are a little more complicated - though the database still works.

Personally I'd use `st.data()` here, and define an `@st.composite` function for strategies that you expect to use in three or more tests.  It's easy to refactor later if you discover you'd like to reuse it.

Zac-HD commented:
Here's how I'd write the complete test:

```python
def _ok_moment_axes(arr):
    ok_axes = []
    with warnings.catch_warnings():
        warnings.simplefilter("error")
        for axis in range(len(arr.shape)):
            with contextlib.suppress(Exception):
                var = stats.moment(arr, moment=2, axis=axis)
                if var > 0:
                    ok_axes.append(axis)
    return ok_axes

class TestStudentTest:
    @given(
        alpha=st.floats(1e-8, 1 - 1e-8),  # one or more bounds implies allow_nan=False
        # alpha=st.floats(0, 1, exclude_min=True, exclude_max=True),  # ??
        arr=npst.arrays(
            dtype=npst.floating_dtypes(),
            shape=npst.array_shapes(min_dims=1, min_side=2),
        ).filter(_ok_moment_axes),
        data=st.data(),  # to choose an ok axis inside the test
    )
    @pytest.mark.parametrize("alternative", ["less", "greater"])
    def test_pvalue_ci(self, data, alpha, arr, alternative):
        # Pick an allowed axis along which to compute
        axis = data.draw(st.sampled_from(_ok_moment_axes(arr)))
        # test the relationship between one-sided p-values and confidence intervals
        res = stats.ttest_1samp(arr, 0, alternative=alternative, axis=axis)
        l, u = res.confidence_interval(confidence_level=alpha)
        popmean = np.expand_dims(l if alternative == "greater" else u, axis=axis)
        res = stats.ttest_1samp(arr, popmean, alternative=alternative, axis=axis)
        np.testing.assert_allclose(res.pvalue, 1 - alpha)
```

mdhaber commented:
Thanks. We'd certainly reuse this (e.g. with the other t-test functions), so I might try `composite`. 

> you've hit the complicated kind of data-generation pretty early.

Yeah, I think that will be common in stats. gh-18930 is a much more straightforward use case. 

Thanks again!

mdhaber commented:
Mmm actually I'm still getting the health check with that code.

Zac-HD commented:
I suspect then that the problem is Hypothesis generating array elements which are quite likely to trigger precision-loss warnings (etc).  In this case, adding `min_value=..., max_value=...` to the `elements` dict should resolve the problem?

mdhaber commented:
Thanks. In `npst.arays` I tried adding:
```python3
elements=dict(min_value=1e-3, max_value=1e3, allow_nan=False,  allow_infinity=False)
```
but then I got
```
hypothesis.errors.InvalidArgument: min_value=0.001 cannot be exactly represented as a float of width 16 - use min_value=0.0010004043579101562 instead.
```
Is that supposed to happen? I thought passing as a `dict` was supposed to avoid trouble due to different data types.

However, adding `elements=dict(allow_nan=False,  allow_infinity=False)` was enough to stop getting the health check. 

Thanks again for your time here. Hopefully I'll get the hang of it and pass it along to other SciPy developers : )

mdhaber commented:
OK, how about:

```python3
    @staticmethod
    @strategies.composite
    def pvalue_ci_data_axis_strategy(draw):
        # Hypothesis strategy to generate an array and corresponding axis
        # for which variance is well defined and nonzero

        # draw an array under shape and value constraints
        dtype = npst.floating_dtypes()
        elements = dict(allow_nan=False, allow_infinity=False)
        shape = npst.array_shapes(min_dims=1, min_side=2)
        data = draw(npst.arrays(dtype=dtype, elements=elements, shape=shape))

        # determine axes for which variance is well-defined and nonzero
        ok_axes = []
        with warnings.catch_warnings():
            warnings.simplefilter("error")
            for axis in range(len(data.shape)):
                with contextlib.suppress(RuntimeWarning):
                    var = stats.moment(data, moment=2, axis=axis)
                    if np.all(var > 0):
                        ok_axes.append(axis)
        assume(ok_axes)  # if there are no valid axes, let hypothesis know

        # draw one of the valid axes
        axis = draw(strategies.sampled_from(ok_axes))

        return data, axis

    @given(alpha=strategies.floats(1e-15, 1-1e-15),
           data_axis=pvalue_ci_data_axis_strategy())
    @pytest.mark.parametrize('alternative', ['less', 'greater'])
    def test_pvalue_ci(self, alpha, data_axis, alternative):
        # test the relationship between one-sided p-values and confidence intervals
        data, axis = data_axis
        # perform test
```

Zac-HD commented:
Ah, that's a tricky interaction!  We want to warn people when their endpoint is not representable in the given precision, and the value of `1e-3` in flaot64 is not exactly representable in float16.  The solution in this case is something like `min_value=float(np.float16(1e-3))` (or copy the value shown), although it's pretty ugly 😢 

Plausibly we should change Hypothesis to automatically adjust the endpoints when the dtype is drawn from a strategy 🤔 

------

And yep, that composite strategy looks great to me!

mdhaber commented:
Great. Just one more thing, hopefully.

I tried drawing examples from my strategy and noticed that all the examples are 1D. When I set `min_dims=2`, it can generate examples (after a short wait), but I wonder whether it is actually generating any cases with multiple dimensions when `min_dims=1`.

Is it right that new cases are generated in each CI run (as long as no counterexample is found) so that eventually it will generate examples with multiple dimensions?

Zac-HD commented:
In general, I just trust that Hypothesis will eventually generate sufficient diverse data to make my test fail - and since it does reroll on every run, it's very very likely that it eventually will.  However, since you're _not_ using `st.data()` it's [easy to ensure that a few specific examples are tested every single time](https://hypothesis.readthedocs.io/en/latest/reproducing.html#providing-explicit-examples), which can make the overall process much faster in expectation:

```python
    @example(alpha=0.5, data_axis=(np.array(...), 1))  # try this example before generating more
    @example(alpha=0.9, data_axis=(np.array(...), 2))  # and this one
    @example(alpha=2.5, data_axis=(np.array(...), 0)).xfail()  # check that invalid inputs fail as expected
    @given(alpha=strategies.floats(1e-15, 1-1e-15),
           data_axis=pvalue_ci_data_axis_strategy())
    @pytest.mark.parametrize('alternative', ['less', 'greater'])
    def test_pvalue_ci(self, alpha, data_axis, alternative):
```

(and I hope that's enough for a while, because I'm about to go offline for a 12-day rafting trip!  seeya late next week 👋)

mdhaber commented:
Awesome! Have a great trip!

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -11,6 +11,10 @@
 from itertools import product
+import hypothesis.extra.numpy as npst
+from hypothesis import given, strategies, assume
+import hypothesis
```
```suggestion
```

tylerjereddy reviewed (commented):
Moved comment over to gh-18930.

tylerjereddy commented:
Oops, wrong PR! I'll move the comment over!

mdhaber commented:
Almost there, I think. 

[0fd30bb](https://github.com/scipy/scipy/pull/18927/commits/0fd30bbdcc92773a5b978b0c0f7282ccd52a4e72) seemed to be needed because different platforms handle warnings differently. Oh well.

Of the two remaining failures, one is still due to Hypothesis not being installed. Suggestions appreciated.

Not sure about the other one. It looks like it might have been a timeout that occured once but not a second time? If that's the case, can we tell Hypothesis to ignore it? That can't have been the fault of `ttest_1samp` or the test.

rgommers commented:
> Of the two remaining failures, one is still due to Hypothesis not being installed. Suggestions appreciated.

Add the `hypothesis` install to the job? It's line 213 on `.github/workflows/windows.yml`.

Kai-Striega commented:
This looks excellent. Thanks, @mdhaber!

> Not sure about the other one. It looks like it might have been a timeout that occured once but not a second time? If that's the case, can we tell Hypothesis to ignore it? That can't have been the fault of ttest_1samp or the test.

There is a setting to [suppress health checks](https://hypothesis.readthedocs.io/en/latest/settings.html#hypothesis.settings.suppress_health_check), maybe that's what you are looking for?

mdhaber commented:
Thanks @Kai-Striega. I've gotten the health check warnings before, and I think that is different. Good to point out that this option exists, but here I made some extra effort to write the tests to avoid that. Fortunately, the timeout issue didn't reoccur. Based on that, I imagine it was something weird on the server rather than a Hypothesis thing, so I don't thing we need to do absolute about it.

rgommers commented:
@mdhaber first thoughts:
- could you look at numpy's `conftest.py` and review/copy how it deals with caching and determinism for `hypothesis`?
- it would be useful to add to the PR summary what the runtime of these tests is when they pass and when they fail

mdhaber commented:
Will commit changes to `conftest.py` shortly.

Added the pass time (~3s) to the top post. I would need to mark it `xslow` before merge. The time when it fails varies widely when it's trying to find a reproducible example. I've seen it take just a few seconds, and I've also seen it take minutes.

rgommers commented:
Thanks Matt. `xslow` seems like the right choice indeed.

> I've considered getting that PR merged with a custom property-based test, then replacing that with a Hypothesis property-based test in this PR. That's more work, but it would make a better case for the use of Hypothesis. What do you think?

That sounds reasonable to me. To make sure I've understood your point: it's the matrix of array lib x array size x dtype x callable there that yields a space that's easy to express for hypothesis but too large/slow to properly explore with a deterministic pytest parametrization?

mdhaber commented:
> it's the matrix of array lib x array size x dtype x callable there that yields a space that's easy to express for hypothesis but too large/slow to properly explore with a deterministic pytest parametrization?

It very well could be, yes. I do think that's an advantage of Hypothesis - rather than writing an exhaustive combinatorial test that runs every CI run (slow), Hypothesis can explore that space a little at a time over many CI runs (each of them fast).

I'm not sure if that's the case here, but what I can say for sure is:
- I think it's easier to express the space with Hypothesis than with raw pytest.
- Even if we try to express all the important aspects of the space with raw pytest, we may miss some that Hypothesis would cover.
- We don't always *try* to write such careful tests, but having Hypothesis as a tool will encourage us to do so.

rgommers commented:
> Hypothesis can explore that space a little at a time over many CI runs

I think that's not what we want. CI runs should be deterministic, otherwise we're inevitably going to get flaky tests (and we tend to ignore unrelated failures on PRs anyway unless they become persistent/annoying). Instead, the tests can be run in exhaustive mode locally (during development, or simply once in a while).

tupui commented:
> I think that's not what we want. CI runs should be deterministic, otherwise we're inevitably going to get flaky tests (and we tend to ignore unrelated failures on PRs anyway unless they become persistent/annoying). Instead, the tests can be run in exhaustive mode locally (during development, or simply once in a while).

Knowing that this is how Hypothesis is working, we could very well work with that. To not miss issues, we could create an issue automatically if on the job there are errors. I believe Scikit-learn has some automatic issue creation on failed CI.

mdhaber commented:
Based on the contents in NumPy's `conftest.py`, that is how it is working there, right? It's what Hypothesis is intended to do. 

Finding real bugs like this is a feature. Hypothesis provides a wrapper that lets you generate the results locally so the failures are easy to investigate, and I believe once it finds a bug like this it will be repeated on CI runs until it is fixed, so it would not be flaky. False positives would be annoying, but that means that the test has a bug, and those bugs should be fixed, too.

rgommers commented:
> I'm sure ours is not the only project that has been concerned with the idea of a nondeterministic test suite.

Yes indeed. We discussed this extensively for NumPy (with @Zac-HD), see https://github.com/numpy/numpy/pull/15189. That said, we don't make heavy use of `hypothesis` in NumPy, it hasn't really taken off. For something like `scipy.stats` with higher-level functionality, it's probably more applicable that for the more "low-level building block" nature of numpy functions.

> Based on the contents in NumPy's `conftest.py`, that is how it is working there, right? I

The comments there are a little confusing. CI jobs are mixed between `spin test` (like our `dev.py test`) and `pip install .` followed by `pytest --pyargs numpy` or `python -c "import numpy; numpy.test()`. The former has a pytest.ini file, the latter does not.

> Finding real bugs like this is a feature, and it's the only practical way to explore large parameter spaces

Finding real bugs is of course a feature and the goal. The question is where to do that. My view is that it's for devs working on the functionality. If you are working on a distribution, you can tell `hypothesis` to keep running until you come back, then go for dinner, and check if it failed afterwards. This is great, it explores a huge space. What you do not want is for random users/contributors to find those failures, that's effectively always a negative outcome and should be avoided. We know enough about unrelated failures I'd say that we know it's bad to have them, whether in CI on unrelated PRs or in installed packages where tests are run.

rgommers commented:
I'd suggest adding a `--randomize` flag to `dev.py test` or some such thing, to make it easy to enable long-running `hypothesis` tests. 

One further thought: such a flag would equally apply to many `pytest` tests. There's a reason we fix seed of random number generation by default - but for a ton of tests that are property-test-like (e.g., linalg decompositions where we assert the decomposition is valid) that isn't actually needed, they're supposed to pass with any seed and the only reason we fix the seed is to avoid flakes.

rgommers commented:
> You're asking for `derandomize=True` in both profiles, even though NumPy seems to allow one to be randomized? What other changes (e.g. names) do you want for the profile settings here?

More like, we always need to use the one profile with `derandomize=True`, _unless_ non-deterministic search is explicitly enabled (so no need to keep the `pytest.ini` check bit).



> I can't do work on `dev.py` until it supports editable installs. In the meantime, can we set whether `derandomize` is True or False based on an environment variable? The environment variable would not be present on CI machines, so it would be default to `derandomize=True`, and developers would opt in by setting the the environment variable, which would ensure that `derandomize=False`.

That sounds good to me, thanks.

tupui commented:
If we use an env variable, can we group it with the other one we have for Array API? More specifically I am talking about this dict:

https://github.com/scipy/scipy/blob/fe877e2f5736cca82ed7b80496ccfb3c060c44fa/scipy/_lib/_array_api.py#L26-L29

mdhaber commented:
For now, I made it independent, since this is not inherently related to the array API.
Is this what you had in mind, Ralf?

mdhaber commented on .gitignore:
```diff
@@ -105,6 +105,10 @@ build-install/*
 .cache/*
 .pytest_cache/*
 
+# hypothesis cache #
+####################
+**/.hypothesis
+
```
Can/should be removed because the Hypothesis home directory is set by the profile to be a temporary directory?

tupui reviewed (commented):
This looks good 👍 Before getting this in, can you add some documentation. That will also help with the communication on the mailing list as people will be able to parse the doc for it.

mdhaber commented:
Added documentation, resolved merge conflicts, added `hypothesis` to new CI jobs, and removed the `xslow` mark (temporarily) to make sure this is still running OK.

mdhaber commented on .github/workflows/linux_meson.yml:
```diff
@@ -56,13 +56,13 @@ jobs:
       run: |
         python -m pip install git+https://github.com/numpy/numpy.git
-        python -m pip install ninja cython pytest pybind11 pytest-xdist pytest-timeout click rich-click doit pydevtool pooch "setuptools<67.3"
+        python -m pip install ninja cython pytest pybind11 pytest-xdist pytest-timeout click rich-click doit pydevtool pooch  hypothesis "setuptools<67.3" 
```
```suggestion
        python -m pip install ninja cython pytest pybind11 pytest-xdist pytest-timeout click rich-click doit pydevtool pooch hypothesis "setuptools<67.3" 
```

mdhaber commented on doc/source/dev/core-dev/distributing.rst.inc:
```diff
@@ -213,3 +214,4 @@ Wheelhouse_, see at the wheel_ and Wheelhouse_ docs.
 .. _manylinux: https://github.com/pypa/manylinux/
 .. _threadpoolctl: https://github.com/joblib/threadpoolctl
 .. _MyST-NB: https://myst-nb.readthedocs.io/
+.. _Hypothesis: https://hypothesis.readthedocs.io/en/latest/index.html
```
```suggestion
.. _Hypothesis: https://hypothesis.readthedocs.io/en/latest/index.html

```

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -3368,6 +3399,21 @@ def test_1samp_ci_iv(self):
         with pytest.raises(ValueError, match=message):
             res.confidence_interval(confidence_level=10)
 
+    @hypothesis.given(alpha=hypothesis.strategies.floats(1e-15, 1-1e-15),
```
```suggestion
    @pytest.mark.xslow
    @hypothesis.given(alpha=hypothesis.strategies.floats(1e-15, 1-1e-15),
```

rgommers commented:
> How does that look, Ralf?
> 
> If the rest looks good, shall I send an email to the mailing list?

LGTM, thanks Matt!

Zac-HD commented on doc/source/dev/contributor/devpy_test.rst:
```diff
@@ -120,6 +120,14 @@ which are disabled even when calling ``python dev.py test -m full``.
 
+By default, tests that use ``Hypothesis`` run with the ``deterministic``
+profile defined in ``scipy/scipy/conftest.py``. This profile includes the
+Hypothesis setting ``derandomize=True`` so the same examples are used until
+Hypothesis, Python, or the test function are updated. To better use
```
It doesn't look like you're pinning Hypothesis (or any other dependencies) when installing in CI, so there will probably be some occasions on which we release a new version - the exact generated examples are explicitly unstable and can change at any time - and the first person to run into that is a new contributor rather than a maintainer.

IMO _just_ pinning Hypothesis would do more harm than that by missing out on new features/bugfixes/perf, but I'd recommend an e.g. `pip-compile` based workflow to pin your transitive dependencies and update them regularly - Hypothesis itself does this with a weekly cronjob.

rgommers replied:
> IMO just pinning Hypothesis would do more harm than that by missing out on new features/bugfixes/perf

Agreed I think

>  the exact generated examples are explicitly unstable and can change at any time 

How often does this happen? Every minor/bugfix release, or only on occasion?

> workflow to pin your transitive dependencies and update them regularly

This is a big job to do well, so I'd prefer not to couple it to this PR unless we really must. And if we do it then I'd much rather go the `conda-lock` route like scikit-learn does, so we can properly control all our dependencies (e.g., testing different BLAS/LAPACK libraries rather than the exact same build of OpenBLAS over and over).

Zac-HD replied:
Thoroughly irregular, but there are changes in the underlying engine a few times per year and a few changes to one or more particular strategies most months.  Of course, any given change is pretty unlikely to make a given test fail!

Agree that pinning deps is a big job and best separated; just wanted to flag it here because when pinned Hypothesis really truly is deterministic.

rgommers replied:
Thanks @Zac-HD. I think we're fine with leaving it unpinned for now then. Given we have very few tests and the chance of breakage is low (it a single change, not a "now things are fully non-deterministic"), we should be fine for quite a while. It's similar to other unpinned deps that may break CI jobs with new releases.

Zac-HD commented on scipy/stats/tests/test_stats.py:
```diff
@@ -3270,6 +3273,34 @@ def test_empty_1d(self):
+                var = stats.moment(data, moment=2, axis=axis)
+                if np.all(var > 0) and np.all(np.isfinite(var)):
+                    ok_axes.append(axis)
+    hypothesis.assume(ok_axes)  # if there are no valid axes, let hypothesis know
```
```suggestion
    hypothesis.assume(ok_axes)  # if there are no valid axes, tell hypothesis to try a different example
```

Zac-HD commented on scipy/conftest.py:
```diff
@@ -153,6 +155,33 @@ def check_fpu_mode(request):
+    os.path.join(tempfile.gettempdir(), ".hypothesis")
+)
+
+# We register two custom profiles for SciPy - for details see
+# https://hypothesis.readthedocs.io/en/latest/settings.html
+# The first is designed for our own CI runs; the latter also
+# forces determinism and is designed for use via scipy.test()
+hypothesis.settings.register_profile(
+    name="nondeterministic", deadline=None, print_blob=True,
+)
+hypothesis.settings.register_profile(
+    name="deterministic",
+    deadline=None, print_blob=True, database=None, derandomize=True,
+    suppress_health_check=list(hypothesis.HealthCheck),
+)
```
It looks like the timing-related flake was from a nondeterministically-exceeded [`deadline`, which is spiritually a healthcheck but uses a separate setting](https://hypothesis.readthedocs.io/en/latest/settings.html#hypothesis.settings.deadline).  Since you're already attending to test runtime, I'd suggest using `deadline=None` to disable it entirely.

Zac-HD reviewed (commented):
:wave: I'm back from the wilderness, and this is looking great!

Aside from the below, https://blog.nelhage.com/post/two-kinds-of-testing/ is my favorite exploration of when to make your tests (non)deterministic and why - it might be useful reading here (along with [pbt is fuzzing](https://blog.nelhage.com/post/property-testing-is-fuzzing/) and [pbt like afl](https://blog.nelhage.com/post/property-testing-like-afl/)).  I mention this because [HypoFuzz](http://pypi.org/project/hypofuzz) exists now, so you can implement exactly the workflow Nelson wanted using Hypothesis!  You'd be fairly early adoptors though, so maybe revisit that in a few months 🙂

mdhaber commented:
[Email](https://mail.python.org/archives/list/scipy-dev@python.org/thread/SPNSRPSR3GFCBJY4QMCCYMZG4J6K3IE6/) sent to mailing list on August 3. I'll resolve the merge conflict and run CI one more time to make sure nothing broke.

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -3271,6 +3274,34 @@ def test_empty_1d(self):
             stats.kurtosis([])
 
 
+@hypothesis.strategies.composite
```
To commit before merge.
```suggestion
@pytest.mark.xslow
@hypothesis.strategies.composite
```

rgommers replied:
This does not seem needed, because this isn't a test function. And the `test_pvalue_ci` it's used from already has the `xslow` marker.

mdhaber replied:
Oops!

mdhaber commented on .gitignore:
```diff
@@ -106,6 +106,10 @@ build-install/*
 .cache/*
 .pytest_cache/*
 
+# hypothesis cache #
+####################
+**/.hypothesis
+
```
I think this can be removed because the Hypothesis home directory is set by the profile to a temp directory.
```suggestion
```

rgommers replied:
indeed

mdhaber reviewed (commented):
Looks like test failures are unrelated. I'll let the reviewer commit the two remaining suggestions (with `[skip ci]`, presumably) before merging, since I wanted someone to confirm the change to `.gitignore`.

mdhaber commented:
@rgommers anything else to do here? Having this in would help with gh-18930 and gh-19023.

mdhaber commented:
I imagine that once `dev.py` can be used to select the profile we can also use it to select the number of tests. In the meantime, editing `max_examples` in `conftest.py` is what I've done.

mdhaber commented on doc/source/dev/contributor/devpy_test.rst:
```diff
@@ -120,6 +120,14 @@ which are disabled even when calling ``python dev.py test -m full``.
+Hypothesis, Python, or the test function are updated. To better use
+Hypothesis' abilities to find counterexamples, select the ``nondeterministic``
+profile by setting the environment variable
+``SCIPY_HYPOTHESIS_PROFILE=nondeterministic`` before running the test suite.
```
```suggestion
``SCIPY_HYPOTHESIS_PROFILE=nondeterministic`` before running the test suite.
The number of examples that are run can be configured by editing the selected
configuration, e.g. adding ``max_examples=100_000``.
```

rgommers reviewed (approved):
Thanks Matt, LGTM now!

> I imagine that once `dev.py` can be used to select the profile we can also use it to select the number of tests.

That sounds like a good solution to me. Can be done in a follow-up PR as needed. Let's get this in now so it can be used in other places/PRs.
